Making experimental processes live

rubenzorrilla opened the pull request:
Adding missing tests for experimental processes and removing use of old analytical function process.

RiccardoRossi committed:
changing name to "experimental" processes to have them to go live.
removing the processes with the same name from the SolidMechanicsApplication (in principle exactly the same input is accepted)

RiccardoRossi committed:
adding missing files

loumalouomega committed:
Revert "Adding an analytical check process"

This reverts commit 895811683cb58005eb0f7541b81e2a14bc2c7934.

loumalouomega committed:
Adding an analytical check process

Scalar case

loumalouomega committed:
Adding analytical check functions

loumalouomega committed:
Adding analytical check functions

loumalouomega committed:
Adding analytical check functions

loumalouomega committed:
Adding analytical check functions

loumalouomega committed:
Adding analytical checks in the tests

loumalouomega committed:
Removing binarys and sporious files

rubenzorrilla committed:
Assign vector by direction process

Renaming experimental_assign_vector_by_direction_process by assign_vector_by_direction_process. Creation of tests for this process.

rubenzorrilla committed:
Supressing use of old apply_custom_function_process

The apply_custom_function_process has been substituted by its updated version.

rubenzorrilla commented:
@RiccardoRossi The old apply_custom_function_process is not used anymore (it was only used in the structural mechanics tests, which now use the "experimental" ones). Shall we delete it?

josep-m-carbonell reviewed (dismissed):
assign_scalar_to_nodes_process.py
assign_vector_components_to_nodes_process.py

can not be deleted

loumalouomega commented:
I think that if there is a replicated process this should disappear with the time. The functionality is very similar, and the adaptation should not be a problem. We can define a period for deprecated processes.

loumalouomega commented:
They are not deprecated, because they add new functionalities and belong to the core, so everyone can use it. I understand that there exist derived ones, that's why I proposed a period of migration. The thing is not replicate functionalities, and make functionalities available to everyone.

RiccardoRossi commented:
@josep-m-carbonell 

i updated the "experimental processes" so that they match the interface you were having in the SolidMechanics application, plus offer some other goodies.
I am proposing to rename them so that they can be used as a drop in replacement for the ones you have + for the old ones we were employing.

Unless i did something wrong they should offer exactly the same functionality and behavior as you were seeing, including corner cases as intervals ending in "End"

The goal, as we discussed previously (@KratosMultiphysics/technical-committee), is to have a unique set of basic processes to be used in all of the applications. We also need to ensure  a test coverage to guarantee
that nothing goes broken or is changed silently over time.

Having said this i may obviously have done something wrong in the implementation (it does not help that are no tests in the SolidMechanicsApplication to verify i did not break anything)
, so if you see some inconsistent behaviour please add to the 
"test_processes.py" in the branch a failing test case, so that i can fix it.

other than that if you have any technical issue with the implementation please vocalize it here so that it can be discussed.

RiccardoRossi committed:
Merge remote-tracking branch 'origin' into making-experimentalprocesses-live

pooyan-dadvand commented:
I'm always against duplicating the code and this case is NOT and exception. 

Meanwhile, I see these processes as a generic tool which can be used in several application and again to avoid duplication I wanted to migrate it to the core. As I see @RiccardoRossi complies with your old interface so I don't see any technical reason to have problem with your derived classes but I may miss something. Would you please explain what is the technical problem here?

RiccardoRossi committed:
checking the variable type at the beginning of the script

loumalouomega commented:
@josep-m-carbonell I understand your position, when I was speaking before about remove duplicated code I was speaking in general, I think this must be a principle. Your process have different functionality, the referent to the rotation displacements, I think the proper thing is to not remove them, and modified them in order to be derived from the ones in the core and keeping the particularities. To avoid conflicts the correct should be never import using: 

      from Kratos... import *

I will restore the files with a commentary in order to have a TODO remembering that it should be modified to be a derived process

Additionally, if you have an idea of tests that you want to add to be run I can contribute

RiccardoRossi commented:
I do believe that the solution that is implemented in the new files is more general than the old one.
To me an important advantage of the technique i am employing (which is the one we discussed a few months back) is that it will work also if you add/remove nodes during the time step. This is important for example if the boundary is remeshed.

In any case, as of today i am not aware of when it will fail. If you add a test showing the issue i will try to fix it.

loumalouomega committed:
Restoring SolidMechanics assign processes (WARNING ADDED)

Please derive this processes  from the core

loumalouomega committed:
Merge branch 'making-experimentalprocesses-live' of https://github.com/KratosMultiphysics/Kratos into making-experimentalprocesses-live

loumalouomega commented:
@josep-m-carbonell I restored the files, adding a Warning, the situation should be temporary, and the process in the solidMechanics should derive from the one in the core

loumalouomega commented:
First let's see what the others think about this change

maceligueta commented:
@loumalouomega and @josep-m-carbonell : in my opinion, the prints don't solve anything, but they are worsening the situation. Instead, let's clarify why the derivation from the new processes is not possible in the SolidMechincs application.

@josep-m-carbonell said:
> Technical issues, the algorithm of fixing and unfixing the degrees of freedom, in the SolidMechanics is solved, you never supplied good a solution for that.

and @RiccardoRossi  said:
> If you add a test showing the issue i will try to fix it

This looks like a dead end road unless you two agree about how to proceed. Am I right or am I missing something?

RiccardoRossi committed:
removing a print that was left behind

RiccardoRossi committed:
Merge branch 'making-experimentalprocesses-live' of https://github.com/KratosMultiphysics/Kratos into making-experimentalprocesses-live

pooyan-dadvand commented:
I don't see it as a dead end. Actually it seems to be in the correct way:

@josep-m-carbonell has found an issue with this implementation which does not have tests

@RiccardoRossi  is not understanding the issue so he asks for a test which fails with the error

So following step would be that Josep Maria adds a test to this branch which fails (which should work with master implementation) so @loumalouomega would see the issue and fix it. 

The good point is that the test would remain to ensure the behaviour of the future modifications

RiccardoRossi commented:
i just removed a print that was wrongly left in the process

RiccardoRossi committed:
moving to c++ the case of a function prescribed in a rotated system of coordinates

RiccardoRossi committed:
adding a test for rotated system of coordinates

pooyan-dadvand commented:
I have just reviewed this pull request and took a look at the repository:

- I could not find the real issue caused by this PR to be fixed
- I also checked the repository and there is NO TEST in SolidMechanicsApplication (NOT even the tests directory)
- @josep-m-carbonell: Your comments are not declaring the technical issue that you are concerned about and at least not transferring a clear technical issue to me. 

I would like to ask you if:
- You can clarify the issue. (It is not a bad idea to fill a issue for that)
- Then as responsible of the SolidMechanicsApplication to provide a tests and as long as we are discussing the issue it is helpful to have a test for this part. (You are claiming that the previous solution was working and is broken but without a test it is not verifiable)
-  you can help us in the migration of this process to the core due to its generic nature

maceligueta commented:
Sorry guys, I need a few hours to check that the PFEM applications still work, I will let you know the status and if extra changes are needed for compatibility. Thanks.

RiccardoRossi committed:
making the callback mechanism more modular, and insulating the kratos from the python internals

RiccardoRossi committed:
making the callback mechanism more modular, and insulating the kratos from the python internals

maceligueta commented:
I just tested this branch and several things are broken when trying to run a PFEM Fluid case.
I will try to fix them, but I will probably need the help of @AFranci. I will block the merge for the moment, until these little problems are solved in this branch.

Commit:
Python version checked. TODO is removed.

Commit:
Cherrypick commit of AFranci's bugfix (I cannot run cases without this fix)

RiccardoRossi commented:
Hi,
       could u test if the test_processes.py runs correctly on your system?

i did change the interface of that function, but i thought i fixed it
consistently.

<http://www.cimne.com/>

maceligueta commented:
@RiccardoRossi : The tests run more or less fine. No problem at all. The previous error I commented was my mistake.
The only thing I am getting is the following:
```
FAIL: test_apply_custom_function_process (__main__.TestProcesses)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_processes.py", line 411, in test_apply_custom_function_process
    self.assertEqual(node.GetSolutionStepValue(VELOCITY_Y), (math.sqrt(abs(node.X*node.Y)))/math.sqrt(2))
AssertionError: 1.7888543820009764 != 1.7888543819998317
```

loumalouomega commented:
@maceligueta Don't use assertEqual, use assertAlmostEqual

https://github.com/KratosMultiphysics/Kratos/wiki/How-to-create-unitary-tests

loumalouomega committed:
Adding patch tests to the SolidMechanicsApplication

Now we have something to test the changes in the SolicMechanics

loumalouomega commented:
@RiccardoRossi and @pooyan-dadvand  now we have patch test to check the changes in the SolidMechanics

maceligueta commented:
Will 'assertAlmostEqual' be exact enough? How big is the tolerance?

pooyan-dadvand commented:
Thanks @loumalouomega for the test.

loumalouomega commented:
@maceligueta You can set the tolerance in AssertAlmostEqual -> https://docs.python.org/3.6/library/unittest.html#unittest.TestCase.assertAlmostEqual

josep-m-carbonell commented:
I have seen the tests, it looks really a promising good thing. Thank you @loumalouomega. 

I have deleted all my initial comments and I will restart:

I would invite you to take a look of the processes of the SolidMechanicsApplication. I think you are trying to clone the same functionality in the core related to that processes. I will make a list of things that maybe you can take into account:

1. Some tolerance have to be added in the when cheking the interval, when a sum of doubles is done the result use to have some error in exces or default. 
( 0.01 + 0.001 ) = 0.0110000000000005 or ( 0.01 + 0.001 ) = 0.010999999999995 
That makes the check of the time interval to fail, usually in the last step. If this tolerance is not added, the last step is unfixed and not fixed again. And usually the static cases not converge.

note.- This also concerns to another intervals or time frequencies for other processes. When time is used sometimes you don't get a GiD output at the same step due to that fact.

2. When you apply a constraint to a derived variable, i.e. VELOCITY, but your dof is  DISPLACEMENT. You must account that the dof to be set is DISPLACEMENT, so some associed derived variables must be taken into account.

3. How to fix and unfix taking into account the different order of the processes.

One exposed solution was fixing and unfixing all dofs in the InitializeSolutionStep and FinalizeSolutionStep of all calculation steps. The other solution ( that is implemented in the SolidMechanics, and I would like to keep), uses an auxiliar variable called INTERVAL_END_TIME and some simple algorithm to check if the end of an interval has been reached and the fixity must be applied again. With this you fix only when is needed.

note.- When new nodes are added they must be cloned and the proper information must be transferred to the new entity, i. e. the dofs and the fixity of that dofs. and then assign the new nodes to the proper submodelpart. You can not do it partially expecting that some other process will arrange it on the way.

4. In order to apply distributed loads in a local reference system, I think it is better to perform the rotation in a C++ process. You can see the implementation in the assign_* processes in the SolidMechanicsApplication/custom_processes/assing*  I think they are clean and simple. ( some Tests are needed to check everything working in a correct way always )

  Related to that local axes, I think that a talk with the reponsibles of the GiD interface is needed. To clarify why they are needed and how to set it from the interface. Not mistake and mix these axes with the structural elements local axes as beams, shells... because they are other local axes. I say this because  we are implementing something in kratos that maybe can not be supplied by the interface to fill correctly this field of the json.....

5. In the assignment of loads we must keep the possibility to assign them to the variables of the conditions instead of the node variables. It is already implemented in the processes of the SolidMechanics. Maybe it would be interesting to have them in the core too.

6. Before pushing any change in the python files, the dependencies to other classes derived from these files must be checked and fixed. In the SolidMechanicsApplication the assign* processes are derived one from the other. ( some Tests are also needed to check it )

I think it's enough for now. We can start with this. 


(note.- usually when you write something in capitals, it means that you are shouting (i.e this is NOT an exception), it can be put in bold instead (i.e this is **not** an exception), I think is more kind).

pooyan-dadvand commented:
@josep-m-carbonell Thanks for your collaboration!

Commit:
Switched from assertEqual to assertAlmostEqual following Vicente's suggestion

Commit:
Merged branch making-experimentalprocesses-live into making-experimentalprocesses-live

RiccardoRossi commented:
Hi,
  i'll answer point by point

regarding adding a tolerance i am fine, however one detail is that with the current implementation
the important information is only if a process is within the interval _at the beginning of the time step_. 
If fixing is performed, or variables are applied this is done at the very beginning of the step. Unfixing is ALWAYS done at exiting.

to be sure if i understand correctly, calling "eps" such tolerance we would do

     if( current_time > interval_begin - epsilon)
          do_something

what value should we choose for "eps"? shall we ask to it to the user or choose it automatically?

regarding your second point, this is something specific of the SolidMechanics (and of the structural mechanics). We can easily construct it on the basis of the "simple" ones that go in the core.

the third point should be fixed by the idea of fixing everything on enterning and unfixing it on the exit. Do you see a case in which this does not work?

point 4 - i agree, i am working on this

point 5 - yes i am moving your processes to the core, with minor modifications. One imporant one however is that the assign_scalar_field you have is using internally an openmp loop which i think can not be done when calling the python callback. Please correct me if i am wrong on this

loumalouomega commented:
What I do for the time is interpolate values, see https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_scripts/from_json_check_result_process.py

If the time is close to the correct one the result will vary slightly

RiccardoRossi commented:
@loumalouomega i don't understand your latest comment

loumalouomega commented:
Instead of using a tolerance i create a list and I interpolate values, so if TIME=0.9999999 it will interpolate and give a very close value to TIME=1.00000, this way you don't need to worry if your tolerance is to small

josep-m-carbonell commented:
Answering to @RiccardoRossi 

I doesn't mind if you fix or unfix everything, if you don't put the tolerance in some cases, depending on the precision of the sum, you will not enter correctly in the interval. I used to put a factor to the current time_step as a tolerance, you can check it in the python processes of the SolidMechanicsApplication. (may be you can try with this @loumalouomega )

About fixing and unfixing everything each step, personally I would avoid it if it is not needed, but we have discussed it previously. Both solutions work, having equivalent pros and cons.

About the OpenMP loop you mentioned in point 5, In elements and nodes it is commented explicitly and says that it does not work in parallel. But now I checked the process for the conditions and it is still active there. I have not detected any error, but it must be commented too.

RiccardoRossi committed:
moving to the core the processes that assign conditions.

modularizing the use of custom_functions by moving to c++ the complete implementation of the callback, so that only the string containing the function to be evaluated is passed

RiccardoRossi committed:
resolving a small conflict

RiccardoRossi committed:
making the custom function capable of assessing wether it depends on space or not

josep-m-carbonell commented:
I will be more precise on my previous response:

1.- This is what is used in the SolidMechanicsApplication: I am sure that can be improved. I think that it must be automatic and you don't have to ask the user anything about that.
    
      def IsInsideInterval(self):
        
        current_time = self.model_part.ProcessInfo[KratosMultiphysics.TIME]
        delta_time   = self.model_part.ProcessInfo[KratosMultiphysics.DELTA_TIME]

        #arithmetic floating point tolerance
        tolerance = delta_time * 0.001

        if( current_time >= (self.interval[0] - tolerance) and current_time <= (self.interval[1] + tolerance) ):
            self.interval_ended = False;
            return True
        else:
            return False

josep-m-carbonell commented:
2.- About the second point. It can be only available for the SolidMechanics and derived applications if you consider that it is not general enough. But the merge of this branch should not destroy it. So, I think that needed implementations to keep it functioning must be included in the branch.

3.- About the third point. As I have said previously, different solutions are possible. I find somewhat absurd to fix and unfix things if it is not necessary to do so. Most of the problems will not need it. It seems like we were not able to find a smarter solution for this. I don't know the real cost of this, I have not seen any number when a large problem is run and you have several thousands of steps.

The points 4 and 5 I think that are clear. Just let me add something to point 4:

4.- When you assign a variable to a condition, the condition expects a value definition for each node. It means that the variables to be assigned are the vectors  POINT_LOADS_VECTOR, LINE_LOADS_VECTOR,  SURFACE_LOADS_VECTOR, etc  ... which type is Vector, instead of POINT_LOAD, LINE_LOAD, SURFACE_LOAD, etc....which type is double. They are needed to set distributed loads. Now works fine using this, I think that the merge should not destroy it, needed implementations to keep it working must be included in the branch.

josep-m-carbonell reviewed (changes requested):
review in the conversation

RiccardoRossi committed:
adding a new utility to unify treatment of intervals, and uploading updated versio of callback_utility, which now accepts a string and implements everything in c++

RiccardoRossi commented:
@josep-m-carbonell 

regarding point2, fine i'll do the implementation of the derived class however i need a test to guide me.

regarding point 3, in my opinion the core should go for the general solution even if it is not the fastest possible in every situation.

on point 4, i am really not clear on what should be the input and output. Could u prepare a test showing what is the expected outcome? we can use that as a basis for discussion

RiccardoRossi commented:
just to update my previous statement (and to do a ping to reactivate the thread)
i  wanted to clarify that the behaviour of applying vector variables to a condition is tested on line 527 of the test_processes.py

did i understand wrongly the expected behaviour? if so ... i need that test to be modified to reflect this problem

pooyan-dadvand commented:
@RiccardoRossi you may reference the code with a link:
https://github.com/KratosMultiphysics/Kratos/blob/making-experimentalprocesses-live/kratos/tests/test_processes.py#L527

josep-m-carbonell commented on kratos/tests/test_processes.py:
```diff
@@ -269,87 +319,268 @@ def test_apply_custom_function_process(self):
         
+    def test_assign_scalar_field_to_conditions(self):
+        model_part = ModelPart("Main")
+        model_part.AddNodalSolutionStepVariable(NODAL_H)
```
why NODAL_H is added here?

RiccardoRossi replied:
this can be removed. I'll do it

josep-m-carbonell commented on kratos/tests/test_processes.py:
```diff
@@ -269,87 +319,268 @@ def test_apply_custom_function_process(self):
+                        "Parameters"            : {
+                            "model_part_name":"Main",
+                            "mesh_id": 0,
+                            "variable_name": "INITIAL_STRAIN",
```
INITIAL_STRAIN,  which type of boundary condition is this?...., is it a single SCALAR ? so it will be applied to each condition integration point? so it will be uniform in each condition or distributed? ....I don't understand what do you try to do in this process ?

RiccardoRossi replied:
INITIAL_STRAIN is a variable of type Vector.
we could use any value of type Vectory at this point, however there are not many of those in the Kratos core. Pick one you like more and i'll use that in the test

josep-m-carbonell replied:
it will be applied to each condition integration point? so it will be uniform in each condition or distributed? ....I don't understand what do you try to do in this process ?

RiccardoRossi replied:
imagine you have nnodes in the conditions

if does a resize of INITIAL STRAIN to nnodes
than it computes the value so that INITIAL_STRAIN[0] is evaluated on node 0, etc

josep-m-carbonell commented on kratos/tests/test_processes.py:
```diff
@@ -269,87 +319,268 @@ def test_apply_custom_function_process(self):
+                    {
+                        "python_module"   : "assign_scalar_field_to_conditions_process",
+                        "kratos_module" : "KratosMultiphysics",
+                        "process_name"          : "AssignValueProcess",
```
AssignValueProcess , it means assign a scalar value?

RiccardoRossi replied:
correct, this should be corrected

josep-m-carbonell commented on kratos/tests/test_processes.py:
```diff
@@ -269,87 +319,268 @@ def test_apply_custom_function_process(self):
+                        "process_name"          : "AssignValueProcess",
+                        "Parameters"            : {
+                            "model_part_name":"Main",
+                            "mesh_id": 0,
```
It is really necessary if it is always mesh_id = 0?

RiccardoRossi replied:
no, it is possible to omit it and it will be set automatically to zero (since this is prescribed in the default values)

josep-m-carbonell replied:
must be kept as a parameter, supplied or prescribed?  If it is not used maybe it can be eliminated.

RiccardoRossi replied:
the user does not need to prescribe it. it can be removed from the test example

RiccardoRossi committed:
changing tests according to JM's review

RiccardoRossi commented:
committed the changes to comply with the review comments

RiccardoRossi commented:
all the reviews were addressed. Can we push this forward please?

josep-m-carbonell committed:
Delete deprecated processes, restore not tested files that must be replaced for the ones of the core.

josep-m-carbonell commented:
I have to test if examples work using new Kratos core processes. I have not done it yet. To not break anything you can go ahead without deleting the files of the SolidMechanics ( I made a commit to this situtation) Then I will remove them later on after checking that  is allright.

RiccardoRossi commented:
@josep-m-carbonell , unfortunately the solution you propose would not work, since the files have the same name, so you would not be sure of what is being imported (if the one in the core or the one in your app). If you tell me that you are going to take a look at this, i can hold a little longer...

josep-m-carbonell commented:
I will take a look asap (next week)

josep-m-carbonell committed:
Recovering wrong deletations

josep-m-carbonell commented:
After reviewing the changes, I am sorry to say that, but I can not delete the files in the SolidMechanicsApp. It will break the current functionality of the implemented processes. The files in the core do not reproduce the current and custom needs for the SolidMechanics and derived applications. I suggest to change the names of the new files that have been pushed into the core to keep the stability of the code.
When a solution to keep the functionality of the processes into the SolidMechanicsApp would be found, a progressive process of adding tests and deleting the files will start.

josep-m-carbonell reviewed (commented):
Change names of the kratos core files

RiccardoRossi commented:
could u please add a small test showing what is the expected behaviour? 
without understanding that, it is simply not possible for me to replicate it.

what i mean is to add something like

def test_expected_structural_behaviour(self):
        ...

to the file "test_processes.py", and mark it as an "expected failure". This way i will know that once that does not fail any longer i can push it

josep-m-carbonell commented:
I have planned a set of tests to be added. However, the problem is that SolidMechanicsApp need for specific things that core do not supplies. I can tell you the detected ones (before doing any test).  The process that assigns and fixes velocity or acceleration do not fixes displacement. The processes that assigns a load must know if the value is given by a function to change the imposed variable. If an interval is [0,0] it must impose an initial condition.... 
I have detected diferent implementations for the assignment processes, some of them completely in c++, others mixing python and c++. Using of old processes mixing varable assigment and dofs fixing at the same execution. I have seen a list of constraints in the vector assignation after deciding that this will not be written lik this... and we are going to use "null". The implementation is a little bit unclear, unclean and disordered. 
I have not seen any solution for the overlaping processes. The tests added to the SolidMechanicsApp use the original processes, not the core ones....So I think this pull request is something not planned correctly, with a lot of hurry and still some issues to solve. Maybe an issue to discuss some of the unsolved questions, as well as to discuss the better implementation can be opened.

maceligueta commented:
@josep-m-carbonell :  it would be great if you specify about every deffect that you find, making a list of the wrong processes or methods is a possibility. If we just say that this is 'unclear' or 'unclean' or 'bad' or 'I have detected something in some processes', nobody can reply, and nobody can fix it!

josep-m-carbonell commented:
@maceligueta, I have not used "bad", and I have not been as much "subjective". It seams as you only look at the my "concluding opinion", it was given after exposing what I reviewed. Before that I have specified everything. I don't have to teach anything to anybody. Do you find my other comments superfluous... well ok. Then my review says only... the implementation is unclear...., if it is like this, ignore it and go ahead with what you want to take into consideration.

maceligueta commented:
@josep-m-carbonell : I just wrote asking for more specific descriptions of the problem, not objectivity. I know that you did not use the word 'bad', but it is an example as general as 'unclean', 'unclear' or 'disordered', which you definitely used to describe who knows what lines of code.
You also talk about the 'overlapping processes' which are also undefined. Do you mean processes over overlapping domains? Processes over overlapping intervals? Overlapping dof's?
If details are not given, in my opinion part of the message is useless (other parts are specific enough).
So, my message had the only intention to ask for some clarification. Maybe you don't want to teach, but many of us want to learn.

pooyan-dadvand commented:
Just to remind that this pull request is for adding some functionality to the core and refactoring the apps to use it instead of their internal solution. It is normal if it generates some incompatibility (especially in features specific to an app). 

Saying that I agree with @maceligueta request for precise description of incompatibilities to see can be incorporated to the core version or solid mechanic application should provide it internally. 

So  @josep-m-carbonell  whould you please point out (in the code) the incompatible parts to be added and try to refactor the solid mechanic app in the same branch in order to make it work?

RiccardoRossi commented:
Hello everyone,
   i just want to clarift the current behaviour, since i have the feeling it is not completely clear. This may also serve as a first step towards a documentation.

**== Imposition of components of a vector variable ==**

In the application of Vector variable the user has some options in deciding what is fixed/free and what is to be imposed or not. I'll use "DISPLACEMENT" as variable to make an example. Also (for the moment) i leave out of the discussion the intervals, not to complicate it further

**OPTION 1**
User prescribes:

    {
            "model_part_name"      : "model_part",
            "variable_name"        : "DISPLACEMENT",
            "value"                       : [10.0, null, "x+y"]
     }

- COMPONENT X :   **assigned value**: 10             - **constrained (fixed)**: True
- COMPONENT Y :   **assigned value**: no action   - **constrained (fixed)**: no action
- COMPONENT X :   **assigned value**: "x+y"         - **constrained (fixed)**: True

**OPTION 2**
User prescribes:

    {
            "model_part_name"   : "model_part",
            "variable_name"        : "DISPLACEMENT",
            "value"                       : [10.0, null, "x+y"],
            "constrained"             : false
      }

- COMPONENT X :   **assigned value**: 10             - **constrained (fixed)**: False
- COMPONENT Y :   **assigned value**: no action   - **constrained (fixed)**: no action
- COMPONENT X :   **assigned value**: "x+y"         - **constrained (fixed)**: False

**OPTION 3**
User prescribes:

    {
            "model_part_name"   : "model_part",
            "variable_name"        : "DISPLACEMENT",
            "value"                       : [10.0, null, "x+y"],
            "constrained"             : [false, true, true]
     }

- COMPONENT X :   **assigned value**: 10             - **constrained (fixed)**: False
- COMPONENT Y :   **assigned value**: no action  - **constrained (fixed)**: no action
- COMPONENT X :   **assigned value**: "x+y"         - **constrained (fixed)**: True

**== Handling of Intervals ==**
Intervals are checked using an epsilon before and after the beginning of the interval.
the function to check if the current time is within an interval is as follows:

       bool IsInInterval(double time )
       {
            const double eps = std::max(1e-14*minterval_begin, 1e-30);
            if(time > minterval_begin-eps && time < minterval_end+eps)
                return true;
            else
                return false;
       }

This means that if a user sets

    {
            "model_part_name"   : "model_part",
            "variable_name"        : "DISPLACEMENT",
            "value"                       : [10.0, null, "x+y"],
            "constrained"             : false,
            "interval"                    : [0,0]
      }

then this will be applied once, provided that a time step happens to have **TIME=0.0**.

I appreciate that we may want to add the capability of applying functions  _exactly once_ when TIME exceeds for the first time the interval begin. I think however this capability shall be provided in a different process.

**== Handling of ACCELERATION and VELOCITY ==**
In order to provide the capability of prescribing ACCELERATION and VELOCITY, in the case in which the time scheme solves in terms of displacemnts, it is necessary to do

-  if ACCELERATION is constrained ==> constrain also DISPLACEMENT
-  if VELOCITY is constrained ==> constrain also DISPLACEMENT

i can easily define two new processes to provide such behaviour. let's temporarily call them

- PrescribeAccelerationProcess
- PrescribeVelocityProcess

that would do, right?

RiccardoRossi commented:
I missed  explaining how fixity is done currently, which follows the decision taken by @KratosMultiphysics/technical-committee 

== Handling of Fixity == 
essentially everything is fixed in the ExecuteInitializeSolutionStep and is unfixed in the ExecuteFinalizeSolutionStep. We believe that this is the most general behaviour since it also allows handling changing meshes during the time step, and leaves everything as "untouched" at the end of the step.

the pseudo code looks something like:

     ExecuteInitializeSolutionStep():
          if(interval.IsInInterval(time))
              if(fixing is required)
                   **fix** dof on all the nodes

     ExecuteInitializeSolutionStep():
            if(fixing is required)
                 **free** dof on all the nodes




having said this, before going into the subject of "code quality" i would like to finalize the decision on the functionality. Once the functionality is there and we have the tests written, we can easily improve the 
quality of the existing code

RiccardoRossi committed:
adding a process to prescribe velocity

RiccardoRossi committed:
adding functions to fix displacements when velocity or acceleration are being prescribed. complete with tests

RiccardoRossi commented:
added the processes 

assign_velocity_for_displacement_driven_solutions_process.py
assign_acceleration_for_displacement_driven_solutions_process.py

with the respective tests.

josep-m-carbonell commented:
== Handling of Fixity ==
A decision taken by @RiccardoRossi. Not the technical committee. Where is the issue discussing this ?
== Handling of ACCELERATION and VELOCITY ==
What about ANGULAR quantities, that must impose ROTATION..... you can create more processes if you want, and if they are specific for the SolidMechanics, put it them the core...
== Imposition of components of a vector variable ==
OPTION 3 is useless, we decided to not implement that option. What does it mean fix a "null" value ? It only offers confussion.
....

RiccardoRossi commented:
Dear @josep-m-carbonell 

The handling of fixity was agreed last summer by a vote in the technical committe, so i believe that a decision has been taken.  I shall acknowledge that the vote was 4 to 1 and not by unanimity.
A issue does not exist on this since the decision was taken long before the github era. In any case @KratosMultiphysics/technical-committee please correct me if i got it wrong.

concerning ANGULAR quantities, fine, i see your point. I'll make the function general so that one specifies what variable to impose and what should be fixed.

concerning OPTION3, it is not useless, since it allows to selectively fix/free one of the variables. This is a feature that is important for example in the FluidMechanicsApplication. As you see you can fix/free all the components with a single bool, however you have the extra flexibility of also doing it component by component (strictly optional). We discussed with this even very recently with @pooyan-dadvand and @rubenzorrilla 
If you took the time of reading my response, you will see that the value of "constrained" for a component is ignored in the case in which that component is taken as "null" (no action) 

having said all of this, it would be useful if you try to be a little more constructive/less negative in your answers

josep-m-carbonell commented:
I don't remember to vote on this issue before summer...Maybe I have a big "lapsus", as usual.  I guess that happened the same as you comment in the OPTION3, initially you were not agree with it and we determined a different alternative. However, now your convenience changed, and a partial unilateral decision was taken... I can not be so glad with that.  It only shows that the criterion of some of you is "superior" compared with the criterion of others.

josep-m-carbonell commented:
It is funny, the polling was 4-1..., so usual result. Anyway we did not vote about that issue but I guess that this would be the result if we do vote right now.... So, no way.

RiccardoRossi committed:
add a generic process to set time derivative variables

RiccardoRossi committed:
merging master into the branch to have it up to date

RiccardoRossi commented:
Addressing the comments i removed the "specific" processes and i added a new "generic" one as follows:

      "process_name"          : "AssignTimeDerivativeProcess",
       "Parameters"            : {
               "model_part_name":"Main",
               "variable_name" : "ACCELERATION",
               "variable_to_be_solved_for" : "DISPLACEMENT",
               "value" : ["t",null,"z"],
               "interval" : [3.0,4.0]
         }

This should hopefully address the concerns about generality of the function.

I am open to address **technical** comments on this or on the implementation or on any of the design choices, however i would ask everyone to leave out of the discussion non technical issues.

RiccardoRossi commented:
@loumalouomega this pull request is getting too big (i mean too many files touched).

can i ask you to move your "test examples" to a separate pull request?
on this aspect i also would suggest that it would be cleaner to compare to the exact analytic solution (which is available for a patch test) instead to reference file.

pooyan-dadvand commented:
@RiccardoRossi what is the relation between the new  `AssignTimeDerivativeProcess` and the original `AssignVectorComponentsToNodesProcess` proposed in this pull request? If its an extension to the original one, I would put some effort to reuse the code from `AssignVectorComponentsToNodesProcess`

RiccardoRossi commented:
The  AssignTimeDerivativeProcess uses 100% of the AssignVectorComponentsToNodesProcess code.

Indeed it constructs and executes internally an object of that type with a slight modification of the input parameters (modification which is done in the constructor). 
**Deriving it would be an option here** , however there are some python details about doing that. Once we agree on the functionality we could put some more work on that.

on the top of what the "base class" do, the  AssignTimeDerivativeProcess also fixes the dofs of the variable to be solved for. To understand why this is needed, let's consider the case of a standard structural element written in terms of DISPLACEMENT (i mean that the dfos to be solved for is actually DISPLACEMENT).
If you want to assign a fixed VELOCITY to some nodes, you'll actually have also to tell that DISPLACEMENT is fixed. This is not trivial since there is some logic involved in deciding what is fixed or not fixed within a given time interval. The current object attempts to inherit that decision from the AssignVectorComponentsToNodesProcess.

One important point is that doing this as described you can actually defer the calculation of the relation between VELOCITY and DISPLACEMENT to the "Predict" of the time scheme, thus ensuring that the same process can work transpaently when different time schemes are employed. As you may remember, we discussed this last year in one of the meetings and we agreed on this solution.

loumalouomega commented:
@RiccardoRossi I am comparing in the patch test with an analytical solution. 

Besides, it is now complicated to me to do any online work, I am in the train

RiccardoRossi commented:
@loumalouomega i am suggesting to compute on the flight the analytical solution instead of reading it from a file. If this is already done, then sorry for the noise.

it is not extremely urgent, i am just asking to move such files to a different pull request when you can. Thx

RiccardoRossi committed:
adding the capability of removing values from a Parameters object

RiccardoRossi committed:
making use of python derivation mechanism to reduce code duplication

RiccardoRossi commented:
@pooyan-dadvand i refactored the code to make use of python derivation mechanism. The code should now be ok. Usage remaind the same...

RiccardoRossi committed:
removing from the branch tests for solid mechanics application

RiccardoRossi committed:
removing files that should not be included in this branch

RiccardoRossi committed:
removing some files which should not be in the commit

RiccardoRossi committed:
removing some files

loumalouomega commented:
@RiccardoRossi Why did you removed some tests from the StructuralMEchanicsApplication?

RiccardoRossi commented:
unless i did it wrong, those tests were not removed

they should still be in the master, but not in that branch... 
anyway it may be that i am messing up with the git, if this is the case please correct me

cheers
Riccardo

loumalouomega commented:
Yes, in `removing some files` you are removing correct tests files

RiccardoRossi committed:
merging master into the branch in the attempt of bringing it up to date

RiccardoRossi committed:
changing name to a file to avoid conflicts

RiccardoRossi committed:
changing name to files

RiccardoRossi committed:
Merge branch 'master' into making-experimentalprocesses-live

RiccardoRossi committed:
getting clean version of some files

RiccardoRossi committed:
getting from master some files that should not be touched

RiccardoRossi committed:
Merge branch 'making-experimentalprocesses-live' of https://github.com/KratosMultiphysics/Kratos into making-experimentalprocesses-live

RiccardoRossi committed:
getting clean versino of some files that should not be modified by this branch

RiccardoRossi committed:
removing two files that should not be there

RiccardoRossi committed:
formatting

RiccardoRossi committed:
removing some comments

RiccardoRossi committed:
removing a comment

rubenzorrilla committed:
Updating fluid dynamics processes

Updating the no slip, inlet and outlet processes in the fluid dynamics application to use the latest version of the experimental processes.

rubenzorrilla committed:
Updated artificial compressibility test

rubenzorrilla committed:
Couette with embedded velocity test updated

rubenzorrilla committed:
Embedded Couette test updated

rubenzorrilla committed:
Embedded reservoir test updated

rubenzorrilla committed:
Navier-Stokes condition test updated

rubenzorrilla committed:
Deleting unused file

pooyan-dadvand commented:
@RiccardoRossi I see that you have changed the names of the processes. Are there still in conflict with the ones in SolidMechanics?

RiccardoRossi commented:
@pooyan-dadvand, no, in principle they are not. The change was done exactly to avoid that conflict

pooyan-dadvand committed:
Merge branch 'master' into making-experimentalprocesses-live

pooyan-dadvand commented:
While the final changes avoids incompatibilities I have removed the conflict of merge and will merge this branch when travis gives ok
